tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/Cache.scala

```diff
@@ -0,0 +1,153 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.Stackable
+
+import scala.collection.immutable.HashMap
+
+/**
+ * An abstract cache that assigns `TargetT`-typed values to `SourceT` typed values, while potentially modifying a
+ * `ContextT`-typed context.
+ *
+ * Example: IntValueCache assigns [[ArenaCell]]s to integers, while modifying a [[PureArena]].
+ *
+ * While it extends [[Stackable]], there's no need to implement an actual stack data structure. By assigning each cache
+ * entry a level, effectively a timestamp for when it was added, we can proxy the state of the stack, by filtering out
+ * entries, based on their levels.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+abstract class Cache[ContextT, SourceT, TargetT] extends Stackable {
+
+  type LevelT = Int
+
+  /**
+   * A stack level, see [[Stackable]]
+   */
+  private var _level: Int = 0
+
+  // the base map, tracking the level at which each entry was added.
+  private var _cache: Map[SourceT, (TargetT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without push/pop
+  protected def cache: Map[SourceT, (TargetT, LevelT)] = _cache
+
+  // reverse mapping
+  private var _reverseCache: Map[TargetT, (SourceT, LevelT)] = HashMap()
+  // inherited classes can read, but not modify without push/pop
```

Comment:
push/pop are not primarily responsible for the updates, rather `addToCache` seems to be the main method. Perhaps correct that in the comment?